src/uoapi/course/course_info.py

```diff
@@ -1,87 +1,128 @@
 import sys
-from time import sleep, perf_counter as pf
-import urllib
-import json
-import itertools as it
 
-import requests
 from bs4 import BeautifulSoup
-#import pandas as pd
+from typing import (
+    Callable,
+    List,
+    TypeVar,
+    cast
+)
+
+import requests
 import regex as re
 
 from uoapi.course import patterns as pt
 from uoapi.course import Prereq
 
-#requests.packages.urllib3.util.ssl_.DEFAULT_CIPHERS += ':DES-CBC3-SHA'
-#timetable_url = 'https://web30.uottawa.ca/v3/SITS/timetable/Search.aspx'
+try:
+    from re import (
+        Match
+    )
+except ImportError:
+    # https://github.com/python/cpython/blob/4bb332cfd1f9740b1e31d2d8b8bf1bedca3439ff/Lib/re.py#L272
+    import sre_compile as _sre
+    Match = type(_sre.compile('', 0).match(''))
+
+# requests.packages.urllib3.util.ssl_.DEFAULT_CIPHERS += ':DES-CBC3-SHA'
+# timetable_url = 'https://web30.uottawa.ca/v3/SITS/timetable/Search.aspx'
 
 # Course Info Parameters
 course_url = 'https://catalogue.uottawa.ca/en/courses/'
 
+T = TypeVar('T')
+
 #############################################################################
 # COURSE INFO SCRAPING
 #############################################################################
 
-def _extract_codes(string, return_all = True):
+
+def _extract_codes(string: str, return_all: bool = True):
     '''
-    Returns course codes found in string; 
+    Returns course codes found in string;
     if multiple codes are found and return_all is False, then returns an invalid code
-    Used in get_subjects.ipynb ''' 
-    codes = list({x.group(0) for x in re.finditer(pt.code_re, string)})
-    if return_all or len(codes) == 1:
-        return codes
-    return 'XXX 0000'
+    Used in get_subjects.ipynb '''
+    return __perform_extraction(
+        string,
+        extracting_fn=lambda x: x.group(0),
+        validating_fn=lambda x: return_all or len(x) == 1,
+        default='XXX 0000'
+    )
+
 
 def _extract_credits(string):
     '''
     Searches string for a number of credits/units
     (Assuming the string is the title of a course)
     Used in get_subjects.ipynb
     '''
-    credits = list({int(x.group(0).split(' ')[0].strip('(')) for x in re.finditer(pt.credit_re, string)})
-    if len(credits) == 1:
-        return credits
-    return [0]
+    return __perform_extraction(
+        string,
+        extracting_fn=lambda x: int(
+            cast(str, x.group(0))
+            .split(' ')[0]
+            .strip('(')
+        ),
+        validating_fn=lambda x: len(x) == 1,
+        default=[0]
+    )
+
+
+def __perform_extraction(
+    string: str,
+    extracting_fn: Callable[[Match], T],
+    validating_fn: Callable[[List[T]], bool],
+    default: T,
+):
```

Comment:
Shouldn't this be the return type?
```suggestion
) -> List[T}:
```